feat(s3): upload file bigger than 50GB

fbeauchamp opened the pull request:
### Description

#### chunk size computation
Chunk size is (length ?? maxLength ) / 10 000 . (min 5MB, max 500MB). 

can be computed from disk utilization (for an xva export) or disk size (for a mirror backup)

#### Transfer strategy 

1. direct transfer : non buffering, only for tar like format , only for bucket without object lock  : memory consumption is constant ( not related to file size). Will use at most one chunk in additional  size on disk
2. direct transfer to temporary file : same as precedent + an additional copy step that truncate the padding. memory consumption is constant . Does not work with backup repository encryption (can't truncate file easily)
3. buffering : use 1 chunk of additional memory ,read one chunk, md5 it,  then transfer it => slower than precedent, no temporary file 
4. double buffering : use 2 chunk ; read 1 chunk , start transfering and reading the next chunk => faster than precedent
5. default mode of `lib-storage` 

I think we should use 2 on non object locked, non encrypted  backup repository, and 5 in other. A file configuration should be present to force the fallback to lib-storage

#### To do 

- [ ] implement sensible maxLenght/lengh computation
- [ ] use it to compute PartSize, throw an error if stream is bigger than 10K PartSize
- [ ] use md5 middleware only if Backup repository has object lock 
- [ ] use direct transfer as a fast path  it BR is not encrypted AND not object locked

### Checklist

- Commit
  - Title follows [commit conventions](https://bit.ly/commit-conventions)
  - Reference the relevant issue (`Fixes #007`, `See xoa-support#42`, `See https://...`)
  - If bug fix, add `Introduced by`
- Changelog
  - If visible by XOA users, add changelog entry
  - Update "Packages to release" in `CHANGELOG.unreleased.md`
- PR
  - If UI changes, add screenshots
  - If not finished or not tested, open as _Draft_

fbeauchamp committed:
fix(fs/s3): throw an error if upload >50GB

fbeauchamp committed:
feat(s3): test upload without sdk

julien-f commented on @xen-orchestra/fs/src/s3.js:
```diff
@@ -224,17 +224,36 @@ export default class S3Handler extends RemoteHandlerAbstract {
+        const MAX_SIZE = MAX_PART * PART_SIZE
+        if (this._readCounter === undefined) {
+          this._readCounter = 0
+        }
```
Use a closure (variable outside of the function) and initialize it to `0`.

fbeauchamp committed:
feat: object lock mode need content md5

also Put object part with a prevalculated md5 and size doesn'c consume additionnal memory against presigned + raw upload

julien-f commented on @xen-orchestra/fs/testupload.mjs:
```diff
@@ -0,0 +1,152 @@
+        region = guessAwsRegion(host),
+      } = parse(url)
+    const client = new S3Client({
+        apiVersion: '2006-03-01',
```
Is this really what we want?

julien-f commented on @xen-orchestra/fs/testupload.mjs:
```diff
@@ -0,0 +1,152 @@
+            keepAlive: true,
+          }),
+        }),
+      })
```
I see we are not passing `tls` option, is this necessary in our S3 implementation?

julien-f commented on @xen-orchestra/fs/testupload.mjs:
```diff
@@ -0,0 +1,152 @@
+            headers:{
+                "content-length": CHUNK_SIZE
+            } 
+        })
```
Isn't it possible to upload all parts in a single chunked request?

julien-f replied:
No, an upload is limited 50 GiB.

julien-f commented on @xen-orchestra/fs/testupload.mjs:
```diff
@@ -0,0 +1,152 @@
+        try{
+            while((data = await read(inputStream, Math.min(remainingBytes, maxChunkPartSize))) !== null){
+                remainingBytes  = await write(data, chunkStream, remainingBytes)
+            }
```
Why a loop? Doesn't the first read give you everything?

julien-f replied:
This is for padding, but we should use a `TransformStream` instead for better perf (less buffering).

fbeauchamp committed:
fix(@xen-orchestra/fs: compute md5 only when needed

fbeauchamp committed:
feat(fs/s3): compute md5 only when needed

fbeauchamp committed:
feat(s3): compute sensible chunk size for s3 upload
